src/drivers/serial/mod.rs

```diff
@@ -0,0 +1,150 @@
+use crate::protocol::{read_all_messages, Protocol};
+use anyhow::Result;
+use std::io::{self, Read, Write};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::broadcast;
+use tokio_serial;
+use tracing::*;
+
+use crate::drivers::{Driver, DriverExt, DriverInfo};
+
+pub struct Serial {
+    pub port_name: String,
+    pub baud_rate: u32,
+}
+
+impl Serial {
+    #[instrument(level = "debug")]
+    pub fn new(port_name: &str, baud_rate: u32) -> Self {
+        Self {
+            port_name: port_name.to_string(),
+            baud_rate,
+        }
+    }
+
+    #[instrument(level = "debug", skip(port))]
+    async fn serial_receive_task(
+        mut port: Box<dyn tokio_serial::SerialPort>,
+        hub_sender: Arc<broadcast::Sender<Protocol>>,
+    ) -> Result<()> {
+        let mut buf = vec![0; 1024];
+        let port_name = port.name().unwrap_or("unknown".to_string());
+
+        loop {
+            tokio::time::sleep(tokio::time::Duration::from_micros(1)).await;
+            match port.read(&mut buf) {
```

Comment:
We should try to `use AsyncReadExt;` to access it asynchronously, maybe using a SerialStream instead of this blockage pooling

Reply:
check now